File: jsone/parser.py
from .AST import Primitive, UnaryOp, ContextValue, BinOp, FunctionCall, ValueAccess, Object, List
from collections import namedtuple
import re
from .shared import TemplateError

Token = namedtuple('Token', ['kind', 'value', 'start', 'end'])


class SyntaxError(TemplateError):

    @classmethod
    def unexpected(cls, got):
        return cls('Found {}, expected !, (, +, -, [, false, identifier, null, number, string, '
                   'true, {}'.format(got.value, "{"))


class Parser(object):
    def __init__(self, tokens, source):
        self.tokens = tokens
        self.source = source
        self.current_token = next(self.tokens)
        self.unaryOpTokens = ["-", "+", "!"]
        self.primitivesTokens = ["number", "null", "true", "false", "string"]
        self.operatorsByPriority = [["||"], ["&&"], ["in"], ["==", "!="], ["<", ">", "<=", ">="], ["+", "-"],
                                    ["*", "/"], ["**"]]

    def take_token(self, *kinds):
        if not self.current_token:
            raise SyntaxError('Unexpected end of input')
        if kinds and self.current_token.kind not in kinds:
            raise SyntaxError.unexpected(self.current_token)
        try:
            self.current_token = next(self.tokens)
        except StopIteration:
            self.current_token = None
        except SyntaxError as exc:
            raise exc

    def parse(self, level=0):
        """  expr : logicalAnd (OR logicalAnd)* """
        """  logicalAnd : inStatement (AND inStatement)* """
        """  inStatement : equality (IN equality)*  """
        """  equality : comparison (EQUALITY | INEQUALITY  comparison)* """
        """  comparison : addition (LESS | GREATER | LESSEQUAL | GREATEREQUAL addition)* """
        """  addition : multiplication (PLUS | MINUS multiplication)* """
        """  multiplication : exponentiation (MUL | DIV exponentiation)* """
        """  exponentiation : propertyAccess (EXP exponentiation)* """
        if level == len(self.operatorsByPriority) - 1:
            node = self.parse_property_access()
            token = self.current_token

            while token is not None and token.kind in self.operatorsByPriority[level]:
                self.take_token(token.kind)
                node = BinOp(token, self.parse(level), node)
                token = self.current_token
        else:
            node = self.parse(level + 1)
            token = self.current_token

            while token is not None and token.kind in self.operatorsByPriority[level]:
                self.take_token(token.kind)
                node = BinOp(token, node, self.parse(level + 1))
                token = self.current_token

        return node

    def parse_property_access(self):
        """  propertyAccess : unit (accessWithBrackets (functionCall)? | DOT id)* """
        node = self.parse_unit()
        token = self.current_token
        while token is not None and (token.kind == "[" or token.kind == "."):
            if token.kind == "[":
                node = self.parse_access_with_brackets(node)
                if self.current_token is not None and self.current_token.kind == "(":
                    node = self.parse_function_call(node)
            elif token.kind == ".":
                token = self.current_token
                self.take_token(".")
                right_part = Primitive(self.current_token)
                self.take_token("identifier")
                node = BinOp(token, node, right_part)
            token = self.current_token
        return node

    def parse_unit(self):
        # unit : unaryOp unit | primitives | contextValue (functionCall)? | LPAREN expr RPAREN | list | object
        token = self.current_token
        if self.current_token is None:
            raise SyntaxError('Unexpected end of input')
        node = None

        if token.kind in self.unaryOpTokens:
            self.take_token(token.kind)
            node = UnaryOp(token, self.parse_unit())
        elif token.kind in self.primitivesTokens:
            self.take_token(token.kind)
            node = Primitive(token)
        elif token.kind == "identifier":
            self.take_token(token.kind)
            node = ContextValue(token)
            if self.current_token is not None and self.current_token.kind == "(":
                node = self.parse_function_call(node)
        elif token.kind == "(":
            self.take_token("(")
            node = self.parse()
            self.take_token(")")
        elif token.kind == "[":
            node = self.parse_list()
        elif token.kind == "{":
            node = self.parse_object()

        return node

    def parse_function_call(self, name):
        """functionCall: LPAREN (expr ( COMMA expr)*)? RPAREN"""
        args = []
        token = self.current_token
        self.take_token("(")
        node = self.parse()

        if node is not None:
            args.append(node)
        while self.current_token.kind == ",":
            self.take_token(",")
            node = self.parse()
            args.append(node)
        self.take_token(")")

        node = FunctionCall(token, name, args)

        return node

    def parse_list(self):
        """  list: LSQAREBRAKET (expr (COMMA expr)*)? RSQAREBRAKET """
        arr = []
        token = self.current_token
        self.take_token("[")

        if self.current_token != "]":
            node = self.parse()
            arr.append(node)

            while self.current_token and self.current_token.kind == ",":
                self.take_token(",")
                node = self.parse()
                arr.append(node)

        self.take_token("]")
        node = List(token, arr)
        return node

    def parse_access_with_brackets(self, node):
        """  valueAccess : LSQAREBRAKET expr |(expr? COLON expr?)  RSQAREBRAKET)"""
        left = None
        right = None
        is_interval = False
        token = self.current_token
        self.take_token("[")
        if self.current_token.kind != ":":
            left = self.parse()
        if self.current_token.kind == ":":
            is_interval = True
            self.take_token(":")
        if self.current_token.kind != "]":
            right = self.parse()

        self.take_token("]")
        node = ValueAccess(token, node, is_interval, left, right)

        return node

    def parse_object(self):
        # """   object : LCURLYBRACE ( STR | ID COLON expr (COMMA STR | ID COLON expr)*)?
        # RCURLYBRACE """
        obj = {}
        token = self.current_token
        self.take_token("{")

        while self.current_token.kind == "string" or self.current_token.kind == "identifier":
            key = self.current_token.value
            if self.current_token.kind == "string":
                key = parse_string(key)
            self.take_token(self.current_token.kind)
            self.take_token(":")
            value = self.parse()
            obj[key] = value
            if self.current_token and self.current_token.kind == "}":
                break
            else:
                self.take_token(",")

        self.take_token("}")
        node = Object(token, obj)

        return node


def parse_string(string):
    return string[1:-1]


class Tokenizer(object):
    def __init__(self):
        ignore = '\\s+'
        patterns = {
            'number': '[0-9]+(?:\\.[0-9]+)?',
            'identifier': '[a-zA-Z_][a-zA-Z_0-9]*',
            'string': '\'[^\']*\'|"[^"]*"',
            # avoid matching these as prefixes of identifiers e.g., `insinutations`
            'true': 'true(?![a-zA-Z_0-9])',
            'false': 'false(?![a-zA-Z_0-9])',
            'in': 'in(?![a-zA-Z_0-9])',
            'null': 'null(?![a-zA-Z_0-9])',
        }
        self.tokens = [
            '**', '+', '-', '*', '/', '[', ']', '.', '(', ')', '{', '}', ':', ',',
            '>=', '<=', '<', '>', '==', '!=', '!', '&&', '||', 'true', 'false', 'in',
            'null', 'number', 'identifier', 'string',
        ]
        token_patterns = [
            '({})'.format(patterns.get(t, re.escape(t)))
            for t in self.tokens]
        if ignore:
            token_patterns.append(('(?:{})'.format(ignore)))
        self.token_re = re.compile('^(?:' + '|'.join(token_patterns) + ')')

    def change_tokenizer(self, grammar):
        # build a regular expression to generate a sequence of tokens
        self.tokens = grammar.tokens
        token_patterns = [
            '({})'.format(grammar.patterns.get(t, re.escape(t)))
            for t in self.tokens]
        if grammar.ignore:
            token_patterns.append(('(?:{})'.format(grammar.ignore)))
        self.token_re = re.compile('^(?:' + '|'.join(token_patterns) + ')')

    def generate_tokens(self, source):
        offset = 0
        while True:
            start = offset
            remainder = source[offset:]
            mo = self.token_re.match(remainder)
            if not mo:
                if remainder:
                    raise SyntaxError(
                        "Unexpected input for '{}' at '{}'".format(source, remainder))
                break
            offset += mo.end()

            # figure out which token matched (note that idx is 0-based)
            indexes = list(
                filter(lambda x: x[1] is not None, enumerate(mo.groups())))
            if indexes:
                idx = indexes[0][0]
                yield Token(
                    kind=self.tokens[idx],
                    value=mo.group(idx + 1),  # (mo.group is 1-based)
                    start=start,
                    end=offset)

File: interpreter/parser/parser.go
package parser

import (
	"fmt"
	"strings"
)

type Parser struct {
	source           string
	tokenizer        Tokenizer
	CurrentToken     Token
	unaryOpTokens    []string
	binOpTokens      []string
	primitivesTokens []string
	operators        [][]string
}

func (p *Parser) NewParser(source string, tokenizer Tokenizer, offset int) (err error) {
	p.source = source
	p.tokenizer = tokenizer
	p.CurrentToken, err = p.tokenizer.Next(p.source, offset)
	p.unaryOpTokens = []string{"-", "+", "!"}
	p.primitivesTokens = []string{"number", "null", "true", "false"}
	p.operators = [][]string{{"||"}, {"&&"}, {"in"}, {"==", "!="}, {">", "<", "<=", ">="}, {"+", "-"}, {"*", "/"}, {"**"}}
	return
}

func (p *Parser) takeToken(kinds ...string) error {
	var err error
	if p.CurrentToken.IsEmpty() {
		return SyntaxError{
			Message: "unexpected end of input",
			Source:  p.source,
			Start:   len(p.source) - 1,
			End:     len(p.source),
		}
	}
	if len(kinds) > 0 && !StringsContains(p.CurrentToken.Kind, kinds) {
		return SyntaxError{
			Message:  fmt.Sprintf("unexpected '%s'", p.CurrentToken.Value),
			Source:   p.source,
			Start:    p.CurrentToken.Start,
			End:      p.CurrentToken.End,
			Expected: kinds,
		}
	}
	p.CurrentToken, err = p.tokenizer.Next(p.source, p.CurrentToken.End)
	return err
}

func (p *Parser) Parse(level int) (node IASTNode, err error) {
	var binaryNode BinOp
	var next IASTNode
	if level == len(p.operators)-1 {
		node, err = p.factor()
		if err != nil {
			return nil, err
		}
		token := p.CurrentToken

		for ; token != (Token{}) && StringsContains(token.Kind, p.operators[level]); token = p.CurrentToken {
			err = p.takeToken(token.Kind)
			if err != nil {
				return nil, err
			}
			next, err = p.Parse(level)
			if err != nil {
				return nil, err
			}
			binaryNode.NewNode(token, next, node)
			node = binaryNode
		}
	} else {
		node, err = p.Parse(level + 1)
		if err != nil {
			return nil, err
		}
		token := p.CurrentToken

		for ; token != (Token{}) && StringsContains(token.Kind, p.operators[level]); token = p.CurrentToken {
			err = p.takeToken(token.Kind)
			if err != nil {
				return nil, err
			}
			next, err = p.Parse(level + 1)
			if err != nil {
				return nil, err
			}
			binaryNode.NewNode(token, node, next)
			node = binaryNode
		}
	}

	return
}

func (p *Parser) factor() (node IASTNode, err error) {
	// factor : unaryOp factor | primitives | (string | list | builtin) (valueAccess)? | LPAREN expr RPAREN | object
	var unaryNode UnaryOp
	var primitiveNode ASTNode
	var next IASTNode
	if p.CurrentToken.IsEmpty() {
		return nil, SyntaxError{
			Message: "unexpected end of input",
			Source:  p.source,
			Start:   len(p.source) - 1,
			End:     len(p.source),
		}
	}
	token := p.CurrentToken
	isUnaryOpToken := StringsContains(token.Kind, p.unaryOpTokens)
	isPrimitivesToken := StringsContains(token.Kind, p.primitivesTokens)

	if isUnaryOpToken {
		err = p.takeToken(token.Kind)
		if err != nil {
			return nil, err
		}
		next, err = p.factor()
		if err != nil {
			return nil, err
		}
		unaryNode.NewNode(token, next)
		node = unaryNode
	} else if isPrimitivesToken {
		err = p.takeToken(token.Kind)
		if err != nil {
			return nil, err
		}
		primitiveNode.NewNode(token)
		node = primitiveNode
	} else if token.Kind == "string" {
		err = p.takeToken(token.Kind)
		if err != nil {
			return nil, err
		}
		primitiveNode.NewNode(token)
		node = primitiveNode
		node, err = p.valueAccess(node)
		if err != nil {
			return nil, err
		}
	} else if token.Kind == "(" {
		err = p.takeToken("(")
		if err != nil {
			return nil, err
		}
		node, err = p.Parse(0)
		if err != nil {
			return nil, err
		}
		err = p.takeToken(")")
		if err != nil {
			return nil, err
		}
	} else if token.Kind == "[" {
		node, err = p.list()
		if err != nil {
			return nil, err
		}
		node, err = p.valueAccess(node)
		if err != nil {
			return nil, err
		}
	} else if token.Kind == "{" {
		node, err = p.object()
		if err != nil {
			return nil, err
		}
	} else if token.Kind == "identifier" {
		node, err = p.builtins()
		if err != nil {
			return nil, err
		}
		node, err = p.valueAccess(node)
		if err != nil {
			return nil, err
		}
	}
	return
}

func (p *Parser) builtins() (node IASTNode, err error) {
	//    builtins : (LPAREN (expr ( COMMA expr)*)? RPAREN)? | (DOT ID)*)
	var args []IASTNode
	var token = p.CurrentToken
	var builtinNode Builtin
	var binOpNode BinOp
	var simpleNode ASTNode
	err = p.takeToken("identifier")
	if err != nil {
		return nil, err
	}
	args = nil
	if p.CurrentToken != (Token{}) && p.CurrentToken.Kind == "(" {
		err = p.takeToken("(")
		if err != nil {
			return nil, err
		}
		node, err = p.Parse(0)
		if err != nil {
			return nil, err
		}
		if node != nil {
			args = append(args, node)
		}

		for p.CurrentToken.Kind == "," {
			err = p.takeToken(",")
			if err != nil {
				return nil, err
			}
			node, err = p.Parse(0)
			if err != nil {
				return nil, err
			}
			args = append(args, node)
		}
		err = p.takeToken(")")
		if err != nil {
			return nil, err
		}
	}
	builtinNode.NewNode(token, args)
	node = builtinNode
	token = p.CurrentToken
	for token != (Token{}) && token.Kind == "." {
		err = p.takeToken(".")
		if err != nil {
			return nil, err
		}
		simpleNode.NewNode(p.CurrentToken)
		err = p.takeToken(p.CurrentToken.Kind)
		if err != nil {
			return nil, err
		}
		binOpNode.NewNode(token, node, simpleNode)
		node = binOpNode
		token = p.CurrentToken
	}

	return
}

func (p *Parser) list() (node IASTNode, err error) {
	//    list : LSQAREBRAKET (expr ( COMMA expr)*)? RSQAREBRAKET)
	var list []IASTNode
	var listNode List
	var token = p.CurrentToken
	err = p.takeToken("[")
	if err != nil {
		return nil, err
	}

	if p.CurrentToken.Kind != "]" {
		node, err = p.Parse(0)
		if err != nil {
			return nil, err
		}
		list = append(list, node)

		for p.CurrentToken.Kind == "," {
			err = p.takeToken(",")
			if err != nil {
				return nil, err
			}
			node, err = p.Parse(0)
			if err != nil {
				return nil, err
			}
			list = append(list, node)
		}
	}
	err = p.takeToken("]")
	if err != nil {
		return nil, err
	}
	listNode.NewNode(token, list)
	node = listNode
	return
}

func (p *Parser) valueAccess(node IASTNode) (IASTNode, error) {
	//    valueAccess : LSQAREBRAKET expr |(expr? SEMI expr?)  RSQAREBRAKET)
	var arrayNode ValueAccess
	var left, right IASTNode
	var err error
	token := p.CurrentToken
	isInterval := false

	token = p.CurrentToken
	for token != (Token{}) && token.Kind == "[" {
		err = p.takeToken("[")
		if err != nil {
			return nil, err
		}
		if p.CurrentToken.Kind != ":" {
			left, err = p.Parse(0)
			if err != nil {
				return nil, err
			}
		}
		if p.CurrentToken.Kind == ":" {
			isInterval = true
			err = p.takeToken(":")
			if err != nil {
				return nil, err
			}
			if p.CurrentToken.Kind != "]" {
				right, err = p.Parse(0)
				if err != nil {
					return nil, err
				}
			}
		}

		err = p.takeToken("]")
		if err != nil {
			return nil, err
		}
		arrayNode.NewNode(token, node, isInterval, left, right)
		node = arrayNode
		token = p.CurrentToken
	}
	return node, err
}

func (p *Parser) object() (node IASTNode, err error) {
	//    object : LCURLYBRACE ( STR | ID SEMI expr (COMMA STR | ID SEMI expr)*)? RCURLYBRACE (DOT ID)?
	var objectNode Object
	var binOpNode BinOp
	var simpleNode ASTNode
	var objValue IASTNode
	obj := make(map[string]IASTNode)
	token := p.CurrentToken
	err = p.takeToken("{")
	if err != nil {
		return nil, err
	}

	for p.CurrentToken.Kind == "string" || p.CurrentToken.Kind == "identifier" {
		key := p.CurrentToken.Value
		if p.CurrentToken.Kind == "string" {
			key = parseString(key)
		}
		err = p.takeToken(p.CurrentToken.Kind)
		if err != nil {
			return nil, err
		}
		err = p.takeToken(":")
		if err != nil {
			return nil, err
		}
		objValue, err = p.Parse(0)
		if err != nil {
			return nil, err
		}
		obj[key] = objValue
		if p.CurrentToken.Kind == "}" {
			break
		} else {
			err = p.takeToken(",")
			if err != nil {
				return nil, err
			}
		}
	}
	err = p.takeToken("}")
	if err != nil {
		return nil, err
	}
	objectNode.NewNode(token, obj)
	node = objectNode
	token = p.CurrentToken
	for token != (Token{}) && token.Kind == "." {
		err = p.takeToken(".")
		if err != nil {
			return nil, err
		}
		simpleNode.NewNode(p.CurrentToken)
		err = p.takeToken(p.CurrentToken.Kind)
		if err != nil {
			return nil, err
		}
		binOpNode.NewNode(token, node, simpleNode)
		node = binOpNode
		token = p.CurrentToken
	}
	return
}

func parseString(s string) string {
	return s[1 : len(s)-1]
}

func CreateTokenizer() (tokenizer Tokenizer) {
	tokenizer = *NewTokenizer(`\s+`, strings.Split(
		`** + - * / [ ] . ( ) { } : , >= <= < > == != ! && || true false in null number identifier string`, " ",
	), map[string]string{
		"number":     `[0-9]+(?:\.[0-9]+)?`,
		"identifier": `[a-zA-Z_][a-zA-Z_0-9]*`,
		"string":     `'[^']*'|"[^"]*"`,
		"true":       `true\b`,
		"false":      `false\b`,
		"in":         `in\b`,
		"null":       `null\b`,
	})
	return
}
